### PR TITLE
Fix builder manager issues

### PR DESCRIPTION
**Pull request by life1347**
This PR aims to fix remaining issues in PR https://github.com/fission/fission/pull/308

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/367)

### CODE REVIEW

File: test/tests/test_buildermgr.sh [hunk 1]
```diff
@@ -23,15 +23,41 @@ checkFunctionResponse() {
...
 fission env delete --name python || true
+kubectl --namespace default get packages|grep -v NAME|awk '{print $1}'|xargs -I@ bash -c 'kubectl --namespace default delete packages @'
 
 echo "Creating python env"
 fission env create --name python --image $PYTHON_RUNTIME_IMAGE --builder $PYTHON_BUILDER_IMAGE
 trap "fission env delete --name python" EXIT
 
-echo "Waiting for env builder to catch up"
-sleep 30
+timeout 60s bash -c waitEnvBuilder
```
Comment (soamvasani) [inline test/tests/test_buildermgr.sh]: Suggest 2-3 minutes, in case that cluster gets slow

File: environments/python3/server.py [hunk 2]
```diff
@@ -25,9 +25,11 @@ def loadv2():
     body = request.get_json()
     filepath = body['filepath']
     functionName = body['functionName']
-    # add filepath into syspath for module import
-    sys.path.append(filepath)
-    fn, path, desc = imp.find_module('user', [filepath])
+    # add module directory into syspath for module import
+    dirs = os.listdir(filepath)
+    for moduleDir in dirs:
```
[the remark this answers is not in the document]
Comment (life1347) [reply environments/python3/server.py]: In previous fetcher `archive` function implementation, fetcher ignores the top level directory of deployment package and only archives the content of deployment package. When fetcher downloads & unarchives the package to `/userfunc` the directory structure would be like:

```
/userfunc/
├── user
│   ├── __init__.py
│   ├── build.sh
│   ├── requirements.txt
│   ├── user.py
│   └── <dependency files>
└── user.tmp
```

With this directory structure, python environment can simply append filepath to sys.path and load user function from it.

However, with the latest changes to `archive` function (archive top-level directory directly instead of its content), the directory structure becomes like:

```
/userfunc/
├── user
│   └── python-srcbuild-1507541251-qhnywl-n0herk-tj4tt8
│       ├── __init__.py
│       ├── build.sh
│       ├── requirements.txt
│       ├── user.py
│       └── <dependency files>
└── user.tmp
```

The directory name `python-srcbuild-1507541251-qhnywl-n0herk-tj4tt8` is generated randomly, python environment has no way to know the exactly directory name to add with. And since `python-srcbuild-1507541251-qhnywl-n0herk-tj4tt8` is the only directory under `/userfunc/user`, so here I let python env uses `os.listdir` to get directory name and add it to sys.path. If environment does not add directory to sys.path we may encounter the issue `ImportError: No module named 'user'`.

If we want to avoid this we can change the name `srcPkgFilename` in https://github.com/fission/fission/blob/master/buildermgr/common.go#L84
and the name `deployPkgFilename` in https://github.com/fission/fission/blob/master/builder/builder.go#L99 to fixed pattern instead of generating randomly.

Pros:
1. fetcher always knows the right path to load function since it is given by poolmgr.

Cons:
1. builder need to check the existence of directories before starting the build process, and remove them if necessary. Also that users may not be able to login env builder to check what happened in  directory after the build failure since the directory is removed. (maybe I worry too much, since such use case should never happen?)

What do you think about this? @soamvasani

[the remark this answers is not in the document]
Comment (soamvasani) [reply environments/python3/server.py]: > fetcher ignores the top level directory of deployment package and only archives the content of deployment package.

Sorry, I don't understand. What is the difference between "contents of the package" and "top level directory of the package"?

[the remark this answers is not in the document]
Comment (soamvasani) [reply environments/python3/server.py]: The goal with unique names is to (a) allow prefetching of functions (b) allow multiple functions to be loaded in the same environment. So any solution to the problems you describe should not assume that there's only one function being fetched.

[the remark this answers is not in the document]
Comment (soamvasani) [reply environments/python3/server.py]: Ok, update from offline conversation -- the previous zip file design was better, and having a random name inside the zip file makes things complicated.

File: environments/python3/server.py [hunk 3]
```diff
@@ -25,9 +25,11 @@ def loadv2():
...
+        sys.path.append(os.path.join(filepath, moduleDir))
+    fn, path, desc = imp.find_module('user')
```
Comment (soamvasani) [inline environments/python3/server.py]: remove hard-coded 'user'

**Comment by soamvasani** [comment]
@life1347 What's the status of this PR? Are you updating this to go back to the earlier zip file structure?

**Comment by soamvasani** [comment]
Thanks @life1347. We can use `FunctionSpec.EntryPoint` for this, right?

**Comment by soamvasani** [comment]
Thanks for investigating, @life1347.

The gist sounds good to me. We can use `FunctionSpec.EntryPoint` to specify the entry point into the package. It will have an environment-specific meaning. This value can be passed on into the FunctionLoadRequest.

My only remaining question is what we'll do in the simpler case where the user specifies just one file. In that case `body["filepath"]` would be a file, not a directory, right? We'll have to handle that case too. Once we do that, this is good for merge, I think.

**Comment by soamvasani** [comment]
Reviewed 7 of 13 files at r1, 3 of 4 files at r2, 1 of 2 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[environments/python3/server.py, line 42 at r4](https://reviewable.io:443/reviews/fission/fission/367#-KxWDn_25OHBW5PnTBsx:-KxWDn_7AXwxaUzUu5LP:bjq4gn3) ([raw file](https://github.com/fission/fission/blob/8919fbeaf286ba7314565642980c43bd65cc1364/environments/python3/server.py#L42)):*
> ```Python
>         # value is a 3-element tuple; otherwise, an exception "ImportError"
>         # is raised.
>         fn, path, desc = imp.find_module(moduleName)
> ```

Can you add a path argument to the `find_module` call? That will ensure that we load `moduleName` from the `filepath`, just in case `moduleName` happens to be a name that's present elsewhere in the path too.

Also, can you change the name of `fn`? This variable is not a filename, it's a file object (like a file handle, not a string containing the name of the file). You could call it simply `f`.

---

*[environments/python3/server.py, line 47 at r4](https://reviewable.io:443/reviews/fission/fission/367#-KxWK5-ZpNnkxW9Oq7-7:-KxWK5-ZpNnkxW9Oq7-8:b-2cn4hg) ([raw file](https://github.com/fission/fission/blob/8919fbeaf286ba7314565642980c43bd65cc1364/environments/python3/server.py#L47)):*
> ```Python
>         # Return module object is the load is successful; otherwise, 
>         # an exception is raised.
>         mod = imp.load_module(moduleName, fn, path, desc)
> ```

You have to close `fn`, which is an open file.  You'll need to wrap this in try/finally. See the example here: https://docs.python.org/3/library/imp.html#examples

---

*[fission/function.go, line 109 at r4](https://reviewable.io:443/reviews/fission/fission/367#-KxYEsdHDGVstnUn3zHa:-KxYEsdJbwhfAofH5xD9:bpvhir0) ([raw file](https://github.com/fission/fission/blob/8919fbeaf286ba7314565642980c43bd65cc1364/fission/function.go#L109)):*
> ```Go
> 		pkgSpec.Source = *createArchive(client, srcArchiveName)
> 		// set pending status to package
> 		pkgStatus = fission.BuildStatusPending
> ```

We should only set BuildStatusPending when there is a `srcArchive` but no `deployArchive`. We should never automatically delete the deploy package.

in other words:
```
if len(srcArchiveName) > 0 && len(deployArchiveName) == 0 {
    pkgStatus = fission.BuildStatusPending
} else {
    pkgStatus = fission.BuildStatusNone
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/367)*

**Comment by soamvasani** [comment]
[I'm just trying out this reviewable thing to see if reviews become easier to manage.]

**Comment by life1347** [comment]
Review status: 10 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[environments/python3/server.py, line 42 at r4](https://reviewable.io:443/reviews/fission/fission/367#-KxWDn_25OHBW5PnTBsx:-Ky5CLHOoGOLSwzcKkqD:b-896fix) ([raw file](https://github.com/fission/fission/blob/8919fbeaf286ba7314565642980c43bd65cc1364/environments/python3/server.py#L42)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Can you add a path argument to the `find_module` call? That will ensure that we load `moduleName` from the `filepath`, just in case `moduleName` happens to be a name that's present elsewhere in the path too.

Also, can you change the name of `fn`? This variable is not a filename, it's a file object (like a file handle, not a string containing the name of the file). You could call it simply `f`.
</blockquote></details>

Done.

---

*[environments/python3/server.py, line 47 at r4](https://reviewable.io:443/reviews/fission/fission/367#-KxWK5-ZpNnkxW9Oq7-7:-Ky5CM04Wmlcu_6T76ei:b-896fix) ([raw file](https://github.com/fission/fission/blob/8919fbeaf286ba7314565642980c43bd65cc1364/environments/python3/server.py#L47)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

You have to close `fn`, which is an open file.  You'll need to wrap this in try/finally. See the example here: https://docs.python.org/3/library/imp.html#examples
</blockquote></details>

Done.

---

*[fission/function.go, line 109 at r4](https://reviewable.io:443/reviews/fission/fission/367#-KxYEsdHDGVstnUn3zHa:-Ky5CNeQrZYDhkvRJPxq:b-896fix) ([raw file](https://github.com/fission/fission/blob/8919fbeaf286ba7314565642980c43bd65cc1364/fission/function.go#L109)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

We should only set BuildStatusPending when there is a `srcArchive` but no `deployArchive`. We should never automatically delete the deploy package.

in other words:
```
if len(srcArchiveName) > 0 && len(deployArchiveName) == 0 {
    pkgStatus = fission.BuildStatusPending
} else {
    pkgStatus = fission.BuildStatusNone
}
```


</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/367)*

**Comment by soamvasani** [comment]
:lgtm:

---

Reviewed 1 of 7 files at r5, 2 of 3 files at r6.
Review status: 7 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/367#-:-KyIDmOj-LgaJWzoCKvP:bnfp4nl)*

**Comment by soamvasani** [comment]
Reviewed 5 of 7 files at r5, 1 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/367)*
